sdks/python/apache_beam/transforms/external.py

```diff
@@ -42,6 +46,97 @@
   grpc = None
 # pylint: enable=wrong-import-order, wrong-import-position, ungrouped-imports
 
+DEFAULT_EXPANSION_SERVICE = 'localhost:8097'
+
+
+def iter_urns(coder, context=None):
+  yield coder.to_runner_api_parameter(context)[0]
+  for child in coder._get_component_coders():
+    for urn in iter_urns(child, context):
+      yield urn
+
+
+class External(ptransform.PTransform):
+  """
+  Base class for transforms that produce external transforms.
+  """
+  _urn = None
+  _schema = None
+
+  def __init__(self, expansion_service=None):
+    self.expansion_service = expansion_service or DEFAULT_EXPANSION_SERVICE
+
+  @classmethod
+  def config_value(cls, obj, typehint=None):
+    """
+    Helper to create a ConfigValue with an encoded value.
+    """
+    if typehint is None:
+      coder = registry.get_coder(type(obj))
+    else:
+      coder = registry.get_coder(typehint)
+    return ConfigValue(
+      coder_urn=list(iter_urns(coder)),
+      payload=coder.encode(obj))
+
+  @classmethod
+  def encode_config(cls, config):
+    schema = cls._schema or {}
+    result = {}
+    for k, v in config.items():
+      typehint = schema.get(k)
+      if v is None and (
+          typehint is None or not isinstance(typehint, typehints.Optional)):
+        # make it easy for user to filter None by default
+        continue
+      result[k] = cls.config_value(v, typehint)
```

Comment:
Do we want to strip the optional wrapping here (e.g. so an Optional[int] is encoded just as a raw into or omitted)?

Reply:
yes, I think so.  I didn't realize that Union/Optional caused coders to fall back to FastPrimitive, which is kind of a bummer.